### DISCUSSION TITLE
Is there a version of this that doesn't use zmk-helpers?

### DISCUSSION THREAD

**Discussion by rieje** [urob/zmk-config#68] [Q&A] [answered]
Is there a version of this zmk-config that doesn't use zmk-helpers? I'm not a programmer and I prefer to refer to official zmk docs to make tweaks to the config. 

For example, I'm trying to add bluetooth keys back to my board (I'm using this zmk-config as my base) but I don't know how to put the bluetooth stuff from this [zmk-config](https://github.com/moergo-sc/glove80-zmk-config/blob/main/config/glove80.keymap) that follows the zmk documented devicetree syntax with the `/ { ... };` syntax. I took a brief look at the guide for devicetree configuration syntax and it doesn't mention this as being optional but the base.keymap from this repo does not follow that syntax.

Actually I think I prefer starting from an empty zmk-config and incorporate parts of this zmk-config as opposed to the other way around since I don't yet understand all parts of the config. 

P.S. Unrelated question, is "timerless homerow mods" merely another variant of homerow mod or a better iteration of it? From reading it seems like the latter, but I'm curious why would people still try whatever the typical implementation is (probably straight from the zmk doc) when the problems of homerow mods are already well known and this "timerless" implementation specifically tackles those issues. I'm also curious whether all of this implementation as well as zmk-helpers are something suitable to be merged into mainline zmk and if not what aspects make it more appropriate as a "third-party module/extension".

Much appreciated.

**Comment by urob** [comment, marked as answer]
> Is there a version of this zmk-config that doesn't use zmk-helpers?

I don't know of anyone who ported this config to stock ZMK without the helpers. That said, it would be pretty straightforward to do so by running `pcpp --passthru-unfound-includes` on `base.keymap`, which will expand all the helper macros and create a stock ZMK version. Alas, the resulting file is almost 4 times as large,  which is one of the motivations for using `zmk-helpers`.

> I took a brief look at the guide for devicetree configuration syntax and it doesn't mention this as being optional but the base.keymap from this repo does not follow that syntax.

Actually it does, it just abstracts the devicetree syntax behind the helper macros, which is another motivation for using `zmk-helpers`. See [this post](https://zmk.dev/blog/2023/12/17/nodefree-config) for more on this.

Btw, there's no need to stick to using only the one or the other -- you can freely mix & match between using the helpers and "pure devicetree" syntax. In fact, that's what I am already doing. For instance, this snippet is pure Devicetree syntax: https://github.com/urob/zmk-config/blob/54d710faa22caad4691475c3adc3a9df7dd638e6/config/base.keymap#L33-L46

**Comment by urob** [comment]
> P.S. Is "timerless homerow mods" merely another variant of homerow mod or a better iteration of it? 

Let me focus on the history here as the readme already covers the "what" and "why". Today, "timeless homerow mods" is just a configuration -- i.e.., a particular combination of ZMK options -- that works particular well. 

A point of confusion might be that when I first published the configuration it relied on two features that at the time weren't yet merged into upstream ZMK:

1. The `on-release` option which is crucial for making positional-hold-taps usable for HRMs. This was implemented in [#1423](https://github.com/zmkfirmware/zmk/pull/1423).
2. The `require-prior-idle` option (formerly known as `global-quick-tap`) that is one of the core components behind the timeless strategy.  This was added by @andrewjrae in [#1187](https://github.com/zmkfirmware/zmk/pull/1187) and [#1387](https://github.com/zmkfirmware/zmk/pull/1387).
